Doc/library/typing.rst

```diff
@@ -1230,6 +1232,83 @@ These are not used in annotations. They are building blocks for creating generic
     ``covariant=True`` or ``contravariant=True``.  See :pep:`484` for more
     details.  By default, type variables are invariant.
 
+.. class:: TypeVarTuple
+
+    :class:`Type variable <TypeVar>` tuple.
+
+    Usage::
+
+        Ts = TypeVarTuple('Ts')
+        Shape = TypeVarTuple('Shape')
+
+    Vanilla type variables enable parameterization with a *fixed* number of
+    types. Type variable tuples, on the other hand, allow parameterization with
+    an *arbitrary* number of types by acting like an *arbitrary* number of type
+    variables wrapped in a tuple. For example::
+
+        class Array(Generic[*Shape]):
+            pass
+
+        from typing import NewType
+        Height = NewType('Height', int)
+        Width = NewType('Width', int)
+
+        1d_array: Array[Height] = Array()  # This is fine
+
+    With one type argument, ``Array`` behaves as if ``Shape = (T,)``, where
+    ``T`` is a type variable. That means ``Array`` behaves as if we'd written
+    ``Generic[T]``. Note the use of the unpacking operator ``*`` to bring ``T``
+    out of the tuple and directly into the square brackets.
+
+    ::
+
+        2d_array: Array[Height, Width] = Array()  # This is also fine!
+
+    With two arguments, it's as if ``Shape = (T1, T2)``, where ``T1`` and ``T2``
+    are type variables; as if ``Array`` was defined with ``Generic[T1, T2]``.
```

Comment:
Maybe a better analogy would be with `*args` for functions? A function using `*args` can be *called* with 0, 1, 2, ... arguments. That doesn't mean it behaves as if it was *defined* with 2 arguments `x, y`.

Reply:
Hmm, I tried this:

```
A vanilla type variable enables parameterization with a single type. A type
variable tuple, on the other hand, allow parameterization with an
*arbitrary* number of types by acting like an *arbitrary* number of type
variables wrapped in a tuple. For example::

    Shape = TypeVarTuple('Shape')

    class Array(Generic[*Shape]): ...

    array_1d: Array[int] = Array()       # This is fine
    array_2d: Array[int, int] = Array()  # This is also fine
    array_0d: Array[()] = Array()        # Even this is fine

In the same way that a function defined with ``*args`` can be *called* with
an arbitrary number of arguments, a generic defined with a ``TypeVarTuple``
be be *parameterized* with an arbitrary number of types.
```

But in fact I think the example arrays communicate the behaviour well enough of their own; the second paragraph feels like repetition (also considering that we end up saying something very similar to the first paragraph).